### PR TITLE
[RDY] lsp:open quickfix botright in default lsp callbacks

### LINKED ISSUE #12241: [LSP] - Be a bit more opinionated on where the quickfix window opens.
I've noticed that when you trigger something that utilizes the *Quickfix List* it always seems to open to the bottom right of what you're doing. I think this may be a bit unexpected. Here are a couple examples of what I mean.

If you are using split and you trigger a find references in the left viewport on a symbol, the quickfix list opens on the bottom of the right viewport.

I've also noticed that this will do the same if the buffer isn't a normal buffer. For example, if you're using somthing like [vista.vim](https://github.com/liuchengxu/vista.vim#neovim) as a symbol explorer and you do a find references, it will actually open the quickfix list in the bottom of the Vista window with is a `nofile` buftype.

I encountered this today when I used the new `workspace/symbol`, which opened to the bottom right of where I was, and it threw me off a bit.

If you are in the right viewport though, then the Quickfix List opens in the bottom of that viewport.

Thoughts on maybe making this a bit more opinionated on where it opens? I think opening in the same viewport makes sense, and I also think not opening in a `nofile` buftype also probably makes sense.

Thanks for the great work on the LSP support. It's daily getting better and better!

- `nvim --version`: Latest master

### PR DESCRIPTION
**Pull request by mattbedell**
Opens quickfix using `botright copen` which seems like a reasonable default for the default lsp callbacks that open quickfix windows.

Closes #12241

### CODE REVIEW

**Comment by tjdevries** [comment]
After I merge #12655 we can make this an optional parameter / configuration option for the handlers, so that if people do not like "botright" they can configure. I hope to merge this week (or at least very soon).

I can give you some more info about that if you're interested in making the change.

**Comment by mattBedell** [comment]
thanks @tjdevries, let me know if I'm heading in the right direction with this: https://github.com/tjdevries/neovim/compare/tjdevries/lsp_diagnostics...mattBedell:lsp-handlers-quickfix?expand=1

I haven't had a chance to check out the test coverage yet, so may be missing some, docs as well.

In any case I'll run your parent branch + the above for the rest of the week.

**Comment by tjdevries** [comment]
I'd go for something more like this:

```lua
--@private
--- Configured open quickfix behavior
--@param config (table)
--     - botright:        (default=true)
--         - Open the quickfix list at the bottom and occupy the full width of the vim window
local function _quickfix_open(config)
  config = vim.lsp._with_extend('vim.lsp.handlers._quickfix_open', {
    botright = true,
  }, config)

  if config.botright == true then
    api.nvim_command("botright copen")
  else
    api.nvim_command("copen")
  end
  api.nvim_command("wincmd p")
end

M['textDocument/references'] = function(_, _, result, _, _, config)
  if not result then return end
  util.set_qflist(util.locations_to_items(result))
  _quickfix_open(config)
end

---------
-- Config should happen in the handler itself, where possible.

-- TODO: Document the keys for config with their defaults.
M['textDocument/references'] = function(_, _, result, _, _, config)
  config = vim.lsp._with_extend('vim.lsp.handlers["textDocument/references"]', {
    enter = true
    copen_cmd = 'botright copen',
  }, config)

  if not result then return end
  util.set_qflist(util.locations_to_items(result))

  api.nvim_command(config.copen_cmd)
  if config.enter then
    api.nvim_command("wincmd p")
  end
end

-------
-- If you want to write a helper function, that's fine,
-- but the config should live in each of the handlers.

local _handle_loclist_config = function(config)
  api.nvim_command(config.copen_cmd)
  if config.enter then
    api.nvim_command("wincmd p")
  end
end

M['textDocument/references'] = function(_, _, result, _, _, config)
  config = vim.lsp._with_extend('vim.lsp.handlers["textDocument/references"]', {
    enter = true
    copen_cmd = 'botright copen',
  }, config)

  if not result then return end
  util.set_qflist(util.locations_to_items(result))
  _handle_loclist_config(config)
end
```

**Comment by mattBedell** [comment]
thanks @tjdevries, made some changes based on your feedback:
https://github.com/tjdevries/neovim/compare/tjdevries/lsp_diagnostics...mattBedell:lsp-handlers-quickfix?expand=1

**Comment by tjdevries** [comment]
Merged the other diagnostics PR, so you can rebase off of master and apply the stuff you were just working on! 

Thanks for being flexible.

**Comment by mattBedell** [comment]
done!

File: runtime/lua/vim/lsp/handlers.lua [hunk 1]
```diff
@@ -76,12 +90,27 @@ M['textDocument/publishDiagnostics'] = function(...)
   return require('vim.lsp.diagnostic').on_publish_diagnostics(...)
 end
 
+--- LSP method textDocument/references
+--@param _ (not used)
```
Comment (tjdevries) [inline runtime/lua/vim/lsp/handlers.lua:94]: do not document `_` params, there is no need.

Comment (mattBedell) [reply runtime/lua/vim/lsp/handlers.lua]: fixed

**Comment by mjlbach** [comment]
I made botright the default, I don't think it's worth it to make it configurable, as its a small function users can directly override should they choose, and the prior behavior was not a great default.

**Pull request closed without merging**